torch/_functorch/_aot_autograd/functional_utils.py

```diff
@@ -191,7 +191,79 @@ def has_metadata_mutation(f_arg, arg, *, check_only_storage_mutation: bool):
         return has_metadata_mutation_
 
 
-def gen_alias_from_base(aliased_base_tensor, target_meta_tensor, target_requires_grad):
+def gen_alias_from_base(
+    aliased_base_tensor,
+    target_meta_tensor,
+    target_requires_grad,
+    target_functional_tensor=None,
+):
+    # Patch the correct requires_grad field of the output tensor, depending on whether:
+    # (i) the reconstructed output (out) was came from a tensor that requires grad or not;
+    # and (ii) the concrete returned output does require grad or not.
+    def patch_requires_grad(out):
+        if aliased_base_tensor.requires_grad and not target_requires_grad:
+            out = out.detach()
+        elif not aliased_base_tensor.requires_grad and target_requires_grad:
+            out.requires_grad_(True)
+        return out
+
+    # If provided, use the target functional tensor for replaying the views.
+    #
+    # In summary, we use the fact that FunctionalTensorWrapper saves the view
+    # functions applied to itself (collected during functionalization) so as
+    # to replay them (view functions) on the aliased_base_tensor.
+    if target_functional_tensor is not None:
+        functional_tensor = target_functional_tensor.tensor
+
+        # First, retrieve the actual base tensor.
+        base = torch._functionalize_base(functional_tensor)
+
+        # Then, apply the necessary ViewMeta so that we reach the offset.
+        # These are ViewMeta that should already be applied to aliased_base_tensor.
+        if target_functional_tensor.offset > 0:
+            base = torch._functionalize_apply_view_metas(
+                target_functional_tensor.tensor,
+                base=base,
+                start=0,
+                end=target_functional_tensor.offset,
+            )
+
+        if base.shape == aliased_base_tensor.shape:
+            # Only apply these functions if the new base (aliased_base_tensor) has the
+            # same shape as the old base. The idea is:
+            #
+            #   1. target_functional_tensor represents the functional tensor that corresponds
+            #      to target_meta_tensor output
+            #
+            #   2. In order to make the new base have the same shape as target_functional_tensor,
+            #      the base of the latter (old base) should have the same shape as the new base
+            #      (that's what FunctionalTensorWrapper guarantees with its regenerate_from_base)
+            #
+            #   3. Otherwise, we have no guarantees the output will have the same shape as
+            #      target_meta_tensor
+            try:
```

Comment:
Also - does this PR handle inplace views properly in all cases? E.g., compare these two tests:
```
def f1(x):
    x.t_(1, 0)
    out = x.unsqueeze(0)
    return out
    
def f2(x):
    out = x.unsqueeze(0)
    x.t_(1, 0)
    return out
    
x1 = torch.ones(2, 4)
x2 = torch.ones(2, 4)
# case 1: unsqueeze happens after the transpose
out2 = f1(x1)
# case 2: unsqueeze, but not transpose on the output
out2 = f2(x2)
```

Reply:
FWIW: I think inplace views are pretty rare (in particular, I'd expect dynamic shapes to be much more common that inplace views in the user code).

So I'm ok with either option, but I think we should either:

(1) make sure that inplace views are fully supported in all of their complexity (I think it definitely complicates this PR)

(2) Explicitly not support the view regeneration logic on inputs that have had inplace views and fallback to the old code path (I think this would simplify the PR quite a bit), and instead worry about dynamic shape support first.

I think either option is ok and probably depends on what you think would be better for XLA. LMK what you think!

Reply:
They seem to work as expected.

Reply:
Ok. I think I see your point.
Currently, it works because inplace-view operations are skipped in dynamo:

https://github.com/pytorch/pytorch/blob/ae6f8d923ceb032d3e933507b071284ba7ec218d/torch/_dynamo/variables/tensor.py#L348-L360

Reply:
I think that, for now, I will drop inplace view support. We can add it afterwards, if we ever need it.

Reply:
Explicitly not supporting in-place view operations, so marking as resolved.